service/task.go

```diff
@@ -840,67 +791,238 @@ func (uis *UIServer) taskModify(w http.ResponseWriter, r *http.Request) {
 }
 
 func (uis *UIServer) testLog(w http.ResponseWriter, r *http.Request) {
+	usr := gimlet.GetUser(r.Context())
+	data := logData{User: usr}
 	vars := gimlet.GetVars(r)
+	raw := (r.FormValue("text") == "true") || (r.Header.Get("Content-Type") == "text/plain")
+
 	logId := vars["log_id"]
+	taskID := vars["task_id"]
+	testName := vars["test_name"]
+	taskExecutionsAsString := vars["task_execution"]
+	taskExec, err := strconv.Atoi(taskExecutionsAsString)
+	grip.Warning(err)
 	var (
-		testLog  *model.TestLog
-		err      error
-		taskExec int
+		logReader io.ReadCloser
+		testLog   *model.TestLog
 	)
 
-	if logId != "" { // direct link to a log document by its ID
-		testLog, err = model.FindOneTestLogById(logId)
-		if err != nil {
-			uis.LoggedError(w, r, http.StatusInternalServerError, err)
+	// Check buildlogger logs first.
+	opts := apimodels.GetBuildloggerLogsOptions{
+		BaseURL:       uis.Settings.Cedar.BaseURL,
+		TaskID:        taskID,
+		TestName:      testName,
+		Execution:     taskExec,
+		PrintPriority: !raw,
+	}
+	logReader, err = apimodels.GetBuildloggerLogs(r.Context(), opts)
+	if err == nil {
+		defer func() {
+			grip.Warning(message.WrapError(logReader.Close(), message.Fields{
+				"task_id":   taskID,
+				"test_name": testName,
+				"message":   "failed to close buildlogger log ReadCloser",
+			}))
+		}()
+		data.Buildlogger = make(chan apimodels.LogMessage, 1024)
+		go apimodels.ReadBuildloggerToChan(r.Context(), taskID, logReader, data.Buildlogger)
+	} else {
+		grip.Error(message.WrapError(err, message.Fields{
+			"task_id":   taskID,
+			"test_name": testName,
+			"message":   "problem getting buildlogger logs",
+		}))
+	}
+
+	// If buildlogger fails, fall back to db.
+	if logReader == nil {
```

Comment:
Can we make a separate helper function for the fallback case?

Reply:
I think we should leave it, we do the same for task logs. This code shouldn't be around for much longer anyways.